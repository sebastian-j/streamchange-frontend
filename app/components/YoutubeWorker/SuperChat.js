import React from 'react';
import PropTypes from 'prop-types';
import styled, { keyframes } from 'styled-components';

const fadeIn = keyframes`
  from {
    opacity: 0;
  }

  to {
    opacity: 1;
  }
`;

const OuterFrame = styled.div`
  background-color: #39acff;
  border-radius: 4px;
  box-shadow: 0 2px 2px 0 rgba(0, 0, 0, 0.14), 0 1px 5px 0 rgba(0, 0, 0, 0.12),
    0 3px 1px -2px rgba(0, 0, 0, 0.2);
  font-family: Roboto, sans-serif;
  left: 50%;
  top: 80vh;
  transform: translateX(-50%);
  position: absolute;
  width: 25vw;
  animation: ${fadeIn} 0.5s ease-out;
`;

const TopBar = styled.div`
  background-color: #0094ff;
  border-radius: 4px 4px 0 0;
  font-weight: bold;
`;

const Logo = styled.img`
  border-radius: 50%;
  height: 32px;
  width: 32px;
  margin: 6px;
`;

const MessageContent = styled.span`
  display: block;
  margin: 6px;
`;

const SuperChat = props => (
  <OuterFrame>
    <TopBar>
      <Logo src={props.imageUrl} alt="logo" /> {props.title}
    </TopBar>
    <MessageContent>{props.message}</MessageContent>
  </OuterFrame>
);

SuperChat.propTypes = {
  imageUrl: PropTypes.string.isRequired,
  message: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
};

export default SuperChat;
