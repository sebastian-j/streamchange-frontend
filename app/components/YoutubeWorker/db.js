import Dexie from 'dexie';

const db = new Dexie('Giveaway');
db.version(1).stores({
  users: 'id, imageUrl, title, message, isModerator, isSponsor, isEligible',
  queue: 'id, imageUrl, title, message, addedAt, lastActiveAt',
  messages: '++id, authorId',
  history: '++id, channelId',
});
export default db;
