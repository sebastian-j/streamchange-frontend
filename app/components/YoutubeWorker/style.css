.three-sections {
    background-color: #eeeeee;
    display: flex;
    flex-direction: row;
    height: 95vh;
}
