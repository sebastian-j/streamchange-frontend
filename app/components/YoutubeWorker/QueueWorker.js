import React, { useEffect, useState } from 'react';
import axios from 'axios';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import styled from 'styled-components';

import { API_URL, PRIVILEGED_CHANNELS } from '../../config';
import { changeColor } from '../../containers/StyleProvider/actions';
import {
  deleteQueueItem,
  pushQueueItem,
  updateQueueItem,
} from '../../containers/QueuePage/actions';
import ChatEmbed from '../ChatEmbed';
import QueueColumn from '../../containers/QueuePage/QueueColumn';
import QueueRules from '../../containers/QueuePage/QueueRules';
import SuperChat from './SuperChat';
import db from './db';

const ThreeSections = styled.div`
  background-color: ${props => props.theme.bodyBackground};
  display: flex;
  flex-direction: row;
  height: 95vh;
`;

const QueueWorker = props => {
  const [timer, setTimer] = useState(null);
  const [superChat, setSuperChat] = useState(null);

  const messageProcessor = () => {
    let nextPageToken = localStorage.getItem('nextPageToken');
    if (nextPageToken === null) {
      nextPageToken = ' ';
    }
    axios
      .get(
        `${API_URL}/v4/liveChat/messages?part=snippet,authorDetails&maxResults=200&id=${
          props.videoId
        }&pageToken=${nextPageToken}`,
      )
      .then(res => {
        localStorage.setItem('nextPageToken', res.data.nextPageToken);
        for (let i = 0; i < res.data.items.length; i += 1) {
          const author = {
            id: res.data.items[i].authorDetails.channelId,
            imageUrl: res.data.items[i].authorDetails.profileImageUrl,
            title: res.data.items[i].authorDetails.displayName,
            message: res.data.items[i].snippet.displayMessage,
            addedAt: res.data.items[i].snippet.publishedAt,
            lastActiveAt: res.data.items[i].snippet.publishedAt,
          };
          const isEligible = res.data.items[i].snippet.displayMessage
            .toLowerCase()
            .includes(localStorage.getItem('queue-command').toLowerCase());
          author.message = author.message.replace(
            localStorage.getItem('queue-command'),
            ' ',
          );
          db.queue
            .where('id')
            .equals(author.id)
            .first()
            .then(user => {
              if (user === undefined && isEligible) {
                db.table('queue')
                  .toArray()
                  .then(items => {
                    if (
                      items.length <
                      parseInt(localStorage.getItem('queue-capacity'), 10)
                    ) {
                      props.pushItem(author);
                    }
                  });
              } else if (typeof user !== 'undefined') {
                props.updateItem(author);
              }
            });
          checkResignation(author);
          superChatFeatures(author, res.data.items[i]);
        }
        clearTimeout(timer);
        setTimer(null);
        setTimer(setTimeout(messageProcessor, res.data.pollingIntervalMillis));
      })
      .catch(() => {
        clearTimeout(timer);
        setTimer(null);
        setTimer(setTimeout(messageProcessor, 6000));
      });
  };

  const superChatFeatures = (author, chatMessage) => {
    if (
      PRIVILEGED_CHANNELS.includes(author.id) ||
      chatMessage.authorDetails.isChatOwner
    ) {
      if (author.message.startsWith('!s ')) {
        setSuperChat({
          title: author.title,
          imageUrl: author.imageUrl,
          message: author.message.replace('!s ', ''),
        });
        setTimeout(() => setSuperChat(null), 6000 + author.message.length * 30);
      } else if (author.message.startsWith('!color ')) {
        setSuperChat({
          title: author.title,
          imageUrl: author.imageUrl,
          message: `${author.title} changed color to ${author.message.replace(
            '!color ',
            '',
          )}`,
        });
        props.onColorChange(author.message.replace('!color ', ''));
        setTimeout(() => setSuperChat(null), 10000);
      }
    }
  };

  const checkResignation = author => {
    if (
      localStorage.getItem('gv-abortCommand') !== null &&
      author.message === localStorage.getItem('gv-abortCommand')
    ) {
      props.deleteItem(author.id);
    }
  };

  useEffect(() => {
    messageProcessor();
    clearTimeout(timer);
  }, []);

  return (
    <ThreeSections>
      <QueueColumn />
      <QueueRules />
      <ChatEmbed videoId={props.videoId} />
      {superChat && (
        <SuperChat
          imageUrl={superChat.imageUrl}
          message={superChat.message}
          title={superChat.title}
        />
      )}
    </ThreeSections>
  );
};

QueueWorker.propTypes = {
  deleteItem: PropTypes.func,
  onColorChange: PropTypes.func,
  pushItem: PropTypes.func,
  updateItem: PropTypes.func,
  videoId: PropTypes.string,
};

export function mapDispatchToProps(dispatch) {
  return {
    deleteItem: id => dispatch(deleteQueueItem(id)),
    onColorChange: col => dispatch(changeColor(col)),
    pushItem: item => dispatch(pushQueueItem(item)),
    updateItem: item => dispatch(updateQueueItem(item)),
    dispatch,
  };
}

export default connect(
  null,
  mapDispatchToProps,
)(QueueWorker);
