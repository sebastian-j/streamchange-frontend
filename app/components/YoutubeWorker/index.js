import React from 'react';
import axios from 'axios';
import PropTypes from 'prop-types';
import ChatEmbed from '../ChatEmbed';
import GiveawayRules from '../GiveawayRules';
import UserList from '../UserList';
import db from './db';
import './style.css';

export default class YoutubeWorker extends React.Component {
  constructor(props) {
    super(props);
    this.messageProcessor = this.messageProcessor.bind(this);
  }

  messageProcessor() {
    let nextPageToken = localStorage.getItem('nextPageToken');
    if (nextPageToken === null) {
      nextPageToken = ' ';
    }
    axios
      .get(
        `https://www.googleapis.com/youtube/v3/liveChat/messages?part=snippet,authorDetails&maxResults=200&liveChatId=${
          this.props.liveChatId
        }&pageToken=${nextPageToken}&key=${this.props.apiKey}`,
      )
      .then(res => {
        localStorage.setItem('nextPageToken', res.data.nextPageToken);
        for (let i = 0; i < res.data.items.length; i += 1) {
          const author = {
            id: res.data.items[i].authorDetails.channelId,
            imageUrl: res.data.items[i].authorDetails.profileImageUrl,
            title: res.data.items[i].authorDetails.displayName,
            message: res.data.items[i].snippet.displayMessage,
            isModerator: res.data.items[i].authorDetails.isChatModerator,
            isSponsor: res.data.items[i].authorDetails.isChatSponsor,
            isEligible: res.data.items[i].snippet.displayMessage.includes(
              localStorage.getItem('keyword'),
            ),
          };
          const message = {
            authorId: res.data.items[i].authorDetails.channelId,
            displayText: res.data.items[i].snippet.displayMessage,
            publishedAt: res.data.items[i].snippet.publishedAt,
          };
          db.users
            .where('id')
            .equals(author.id)
            .first()
            .then(user => {
              if (user === undefined) {
                db.table('users').add(author);
              } else {
                db.table('users')
                  .where('id')
                  .equals(author.id)
                  .modify({
                    message: author.isEligible ? author.message : user.message,
                    isEligible:
                      user.isEligible === true ? true : author.isEligible,
                  });
              }
            });
          db.table('messages').add(message);
        }
        setTimeout(this.messageProcessor, res.data.pollingIntervalMillis);
      });
  }

  componentDidMount() {
    this.messageProcessor();
  }

  render() {
    return (
      <div className="three-sections">
        <UserList />
        <GiveawayRules />
        <ChatEmbed videoId={this.props.videoId} />
      </div>
    );
  }
}

YoutubeWorker.propTypes = {
  apiKey: PropTypes.string.isRequired,
  videoId: PropTypes.string,
  liveChatId: PropTypes.string.isRequired,
};
