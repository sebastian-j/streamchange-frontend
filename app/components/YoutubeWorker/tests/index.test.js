import React from 'react';
import renderer from 'react-test-renderer';

import YoutubeWorker from '../index';

describe('<YoutubeWorker />', () => {
  it('should match the snapshot', () => {
    const renderedComponent = renderer
      .create(
        <YoutubeWorker
          apiKey="key"
          channelId="id"
          liveChatId="chatId"
          videoId="vidId"
        />,
      )
      .toJSON();
    expect(renderedComponent).toMatchSnapshot();
  });
});
