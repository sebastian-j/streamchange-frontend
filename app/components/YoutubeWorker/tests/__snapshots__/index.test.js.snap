// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<YoutubeWorker /> should match the snapshot 1`] = `
<div
  className="YoutubeWorker__ThreeSections-dpgts1-0 cjrDdj"
>
  <div
    className="UserList__UserListPanel-sc-7bgn44-0 gjWqbF"
  >
    <h2
      className="PanelTitle-sc-100fqnu-0 dfyaea"
    >
      <span>
        Users
      </span>
    </h2>
    <div
      className="MuiFormControl-root MuiTextField-root StyledTextField-sc-1otd4ny-0 dczFRm MuiFormControl-marginDense MuiFormControl-fullWidth"
    >
      <label
        className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-marginDense"
        data-shrink={false}
      >
        Search
      </label>
      <div
        className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl MuiInputBase-marginDense"
        onClick={[Function]}
      >
        <input
          aria-invalid={false}
          autoFocus={true}
          className="MuiInputBase-input MuiInput-input MuiInputBase-inputMarginDense MuiInput-inputMarginDense"
          disabled={false}
          name="search"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={false}
          type="text"
          value=""
        />
      </div>
    </div>
    <ul
      className="UserList__Ul-sc-7bgn44-1 dFcjyK"
    />
    <button
      className="MuiButtonBase-root MuiButton-root UserList__StyledButton-sc-7bgn44-2 hMUTnL MuiButton-text"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-label"
      >
        <span>
          Clear all
        </span>
      </span>
      <span
        className="MuiTouchRipple-root"
      />
    </button>
    <span
      className="UserList__Counts-sc-7bgn44-3 hhpEOQ"
    >
      <span>
        0 out of 0 are participating
      </span>
    </span>
  </div>
  <div
    className="Panel-sc-15d2ub8-0 hhnpCF"
  >
    <h2
      className="PanelTitle-sc-100fqnu-0 dfyaea"
    >
      <span>
        Giveaway rules
      </span>
    </h2>
    <button
      className="GiveawayRules__UserTypeButton-sc-142hb20-0 bmMAMZ"
      name="forMods"
      onClick={[Function]}
      type="button"
    >
      <span>
        Moderators
      </span>
    </button>
    <button
      className="GiveawayRules__UserTypeButton-sc-142hb20-0 bmMAMZ"
      name="forSponsors"
      onClick={[Function]}
      type="button"
    >
      <span>
        Sponsors
      </span>
    </button>
    <button
      className="GiveawayRules__UserTypeButton-sc-142hb20-0 bmMAMZ"
      name="forRegulars"
      onClick={[Function]}
      type="button"
    >
      <span>
        Regulars
      </span>
    </button>
    <div
      className="MuiFormControl-root MuiTextField-root StyledTextField-sc-1otd4ny-0 dczFRm MuiFormControl-marginDense MuiFormControl-fullWidth"
    >
      <label
        className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-marginDense"
        data-shrink={false}
      >
        Prize (optional)
      </label>
      <div
        className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl MuiInputBase-marginDense"
        onClick={[Function]}
      >
        <input
          aria-invalid={false}
          autoFocus={true}
          className="MuiInputBase-input MuiInput-input MuiInputBase-inputMarginDense MuiInput-inputMarginDense"
          disabled={false}
          name="prize"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={false}
          type="text"
          value={null}
        />
      </div>
    </div>
    <div
      className="MuiFormControl-root MuiTextField-root StyledTextField-sc-1otd4ny-0 dczFRm MuiFormControl-marginDense MuiFormControl-fullWidth"
    >
      <label
        className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-marginDense"
        data-shrink={false}
        htmlFor="keyword"
      >
        Keyword
      </label>
      <div
        className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedEnd MuiInputBase-marginDense"
        onClick={[Function]}
      >
        <input
          aria-invalid={false}
          autoFocus={true}
          className="MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedEnd MuiInputBase-inputMarginDense MuiInput-inputMarginDense"
          disabled={false}
          id="keyword"
          name="keyword"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={false}
          type="text"
          value=""
        />
        <div
          className="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-marginDense"
        >
          <button
            aria-describedby={null}
            aria-label="toggle keyword visibility"
            className="KeywordInput__ToggleVisibilityBtn-sc-1i66nol-0 ldEhFZ"
            name="showKeyword"
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onMouseLeave={[Function]}
            onMouseOver={[Function]}
            onTouchEnd={[Function]}
            onTouchStart={[Function]}
            title="Toggle keyword visibility"
            type="button"
          >
            <svg
              height="24"
              viewBox="0 0 24 24"
              width="24"
            >
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
              <path
                d="M12 4.5C7 4.5 2.73 7.61 1 12c1.73 4.39 6 7.5 11 7.5s9.27-3.11 11-7.5c-1.73-4.39-6-7.5-11-7.5zM12 17c-2.76 0-5-2.24-5-5s2.24-5 5-5 5 2.24 5 5-2.24 5-5 5zm0-8c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3z"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
    <div>
      <div
        className="MuiFormControl-root RaffleWrapper__StyledFormControl-emso03-1 jqDDzA MuiFormControl-marginNormal"
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
          data-shrink={true}
          id="animation-select"
        >
          <span>
            Animation
          </span>
        </label>
        <div
          className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
          labelId="animation-label"
          onClick={[Function]}
        >
          <div
            aria-haspopup="listbox"
            className="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input MuiInputBase-inputSelect"
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={0}
          >
            <span>
              CS:GO case opening
            </span>
          </div>
          <input
            type="hidden"
            value={0}
          />
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root MuiSelect-icon"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
        </div>
      </div>
      <div
        className="NumericInput__Container-sc-2ueyqw-0 iGHMIM"
      >
        <div
          className="NumericInput__Label-sc-2ueyqw-1 fqrMbr"
        >
          Animation duration (seconds)
        </div>
        <button
          className="NumericInput__Button-sc-2ueyqw-2 jSXsrE"
          onClick={[Function]}
          type="button"
        >
          -
        </button>
        <input
          className="NumericInput__NumberDisplay-sc-2ueyqw-3 iputsX"
          onBlur={[Function]}
          onChange={[Function]}
          value={7}
        />
        <button
          className="NumericInput__Button-sc-2ueyqw-2 bqvYTn"
          onClick={[Function]}
          type="button"
        >
          +
        </button>
      </div>
      <button
        className="RaffleWrapper__StartButton-emso03-0 hsCXbs"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        <span>
          Roll it!
        </span>
      </button>
    </div>
  </div>
  <div
    className="Panel-sc-15d2ub8-0 hhnpCF"
  >
    <h2
      className="PanelTitle-sc-100fqnu-0 dfyaea"
    >
      <span>
        Chat
      </span>
    </h2>
    <iframe
      className="chat-frame ChatEmbed__ChatFrame-g1101a-0 bRtIsU"
      src="https://www.youtube.com/live_chat?v=vidId&embed_domain=localhost"
      title="Youtube Chat"
    />
  </div>
</div>
`;
